rfc/indexing-design.md

```diff
@@ -0,0 +1,228 @@
+
+After the add-blob process completes, and the user receives an [accept-blob receipt](https://github.com/w3s-project/specs/blob/main/w3-blob.md#accept-blob-receipt), the user may then choose to make their data publicly queryable by publishing it on [IPNI](https://github.com/ipni/specs/blob/main/IPNI.md). The user optionally invokes an [index-add](https://github.com/w3s-project/specs/blob/main/w3-index.md#index-add) capability to publish the blob-index multihashes to IPNI so that they can be used to look up location commitments and retrieve blob-index information and blob data. See [W3 Index](https://github.com/w3s-project/specs/blob/main/w3-index.md#w3-index) for more.
+
+After publishing to IPNI, any user looking for content can query IPNI for the blob-index of that content. The user read the blob-index to determine what data shard needs to be retrieved, and then asks IPNI for the location of that shard.
+
+Finally the user uses the shard location, with range information from the blob index, to retrieve a specific piece of data from w3s.
+
+## Index Query
+
+A w3s client can query indexed information by specific attributes, and can filter retrieved query results.
+
+The queryable attributes are as follows.
+- Site: the location(s) where all blob data is available
+- Shard: the location(s) where a specific shard of data is available
+- Slice: the location(s) within a shard where a block of data is available
+
+Query results can be filtered to select only results that the user has access to, or for data supplied by specific data providers.
+
+### Query Specification
+
+```
+match = {
+  subject=[myID, myOrgID],
+  location=spBaseURL,
+}
+results = Query(multihash="bafy...", match)
+```
+
+### Query Operation Diagram
+
+![w3up-query-diagram](w3up-query-diagram.png)
+
+## Indexing Components
+
+The indexing components are intended for deployment on the w3up gateway. They may also be deployed to the w3up client, where they will not be shared with other clients.
+
+### W3up Index
+
+The w3up index layer is responsible for reading a query specification, and then
+1. Creating a set of IPNI queries needed to get the requested index content
+2. Sending the queries to the w3up IPNI cache
+3. Filtering returned results
+4. Presenting results to the query client.
+
+### W3up IPNI Cache
+
+The w3up IPNI Cache is a cache that holds IPNI query results. The cache is:
+- Limited-size LRU. Discards the least recently used data when at storage capacity.
+- Temporary. Evicts items that have been cached for longer than the cache expiration time, regardless of last access. This allows changes to query results to be seen.
+- Negative (empty) result-aware. Caching empty results prevents more expensive IPNI queries for data that is not indexed.
+
+This is a read-through cache, meaning if it does not hold the request query results, then it forwards the query on to IPNI, and caches the results. This includes caching an empty response.
```

Comment:
I think caching an empty response can be problematic as we may not have a record on first query which would get cached and prevent us from discovering records on subsequent queries. Perhaps short TTL on empty result could be a way to address this or some form of exponential backoff.

Reply:
Hopefully addressed by read-on-write -- if we're capturing negative responses we need to write the cache when we publish IPNI.

Because otherwise @Gozala makes a very good point and we've seen this exact problem with IPNI and Cloudflare caching -- 404s getting cached for 10 minutes if you query too soon.

Reply:
Having cache-on-write fixes the case where new data is stored subsequent to a query for that data. When the new data is stored, or location info is updated, that would remove any negative cache. Without cache-on-write then a short TTL is necessary.  With cache-on-write, is there still a reason for a short TTL?

Negative cache entries should be kept in a separate cache so that a negative entry cannot evict a positive entry due to LRU, thereby removing the more valuable positive entries and allowing a misbehaving client to empty the cache.

Will clarify in doc.